php/ext/google/protobuf/protobuf.c

```diff
@@ -242,13 +242,18 @@ bool ObjCache_Get(const void *upb_obj, zval *val) {
 // -----------------------------------------------------------------------------
 
 void NameMap_AddMessage(const upb_MessageDef *m) {
-  char *k = GetPhpClassname(upb_MessageDef_File(m), upb_MessageDef_FullName(m));
+  char *k = GetPhpClassname(upb_MessageDef_File(m), upb_MessageDef_FullName(m), false);
   zend_hash_str_add_ptr(&PROTOBUF_G(name_msg_cache), k, strlen(k), (void*)m);
+  char *k2 = GetPhpClassname(upb_MessageDef_File(m), upb_MessageDef_FullName(m), true);
+  if (strcmp(k, k2) != 0) {
```

Comment:
We need to be careful about the CPU costs of what we are doing here. The cost of loading large schemas is significant, and this is in the critical path for important users like Ads.

As written this will always generate the name twice, compare them, and throw away one if they are the same. Could we change it to only generate the alternate name if the primary name is a reserved word?

Reply:
The problem is by the time we can check if the name is reserved, we're way deep in the `is_reserved` function, and then we lose the ability to compare once the full PHP classname is returned.

Reply:
I refactored it to do as I mention above (read the ending classname and return `true` if it was previously unreserved)

I think there's probably a better way and I'm definitely open to ideas.